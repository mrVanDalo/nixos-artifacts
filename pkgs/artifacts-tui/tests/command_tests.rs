use insta_cmd::{Spawn, SpawnExt, StdinCommand, assert_cmd_snapshot, get_cargo_bin};
use std::path::{Path, PathBuf};
use std::process::Command;

fn cli() -> Command {
    Command::new(get_cargo_bin("artifacts-cli"))
}

fn sdtin_cli(stdin: &str) -> StdinCommand {
    StdinCommand::new(get_cargo_bin("artifacts-cli"), stdin)
}

fn project_root() -> PathBuf {
    // tests run with CWD at crate root, but compute robustly
    PathBuf::from(env!("CARGO_MANIFEST_DIR"))
}

/// Helper to manage a fixed temp directory for deterministic snapshots.
/// It ensures the directory exists before a test, and after the test
/// verifies it is empty; if not, the test fails, but the directory is
/// deleted in any case.
struct TempTestEnv {
    path: PathBuf,
}

impl TempTestEnv {
    fn new() -> Self {
        let path = PathBuf::from("/tmp/artifacts-tui-ci");
        let _ = std::fs::remove_dir_all(&path);
        std::fs::create_dir_all(&path).expect("failed to create fixed tmp dir");
        TempTestEnv { path }
    }

    fn apply_env<'a>(&self, cmd: &'a mut StdinCommand) -> &'a mut StdinCommand {
        // StdinCommand::env returns &mut Command; we don't need the return value here.
        let _ = cmd.env("TMPDIR", &self.path);
        let _ = cmd.env("ARTIFACTS_TUI_TEST_FIXED_TMP", "1");
        cmd
    }

    fn is_empty_dir(path: &Path) -> std::io::Result<bool> {
        let mut entries = std::fs::read_dir(path)?;
        Ok(entries.next().is_none())
    }

    /// Check emptiness, then delete directory. Returns Ok(()) if empty,
    /// Err(msg) if not empty. In both cases the directory is removed.
    fn finish(self) -> Result<(), String> {
        let empty = match Self::is_empty_dir(&self.path) {
            Ok(v) => v,
            Err(e) => return Err(format!("failed to read tmp dir: {}", e)),
        };
        // Always attempt deletion
        if let Err(e) = std::fs::remove_dir_all(&self.path) {
            return Err(format!("failed to remove tmp dir: {}", e));
        }
        if empty {
            Ok(())
        } else {
            Err("temporary directory not empty before cleanup".to_string())
        }
    }
}

#[test]
fn scenario_simple() {
    let root = project_root();
    let backend = root.join("examples/scenario_simple/backend.toml");
    let make = root.join("examples/scenario_simple/make.json");

    let env = TempTestEnv::new();

    let mut cmd = sdtin_cli("one\ntwo\n");

    env.apply_env(&mut cmd)
        .arg("generate")
        .arg(backend)
        .arg(make)
        .pass_stdin("one")
        .pass_stdin("two");

    // Verify and cleanup
    env.finish().expect("temp folder not empty at end of test");

    assert_cmd_snapshot!(cmd);
}

#[test]
fn generator_missing_scenario() {
    let root = project_root();
    let backend = root.join("examples/generator_missing/backend.toml");
    let make = root.join("examples/generator_missing/make.json");

    let env = TempTestEnv::new();

    let mut cmd = sdtin_cli("one\ntwo\n");

    env.apply_env(&mut cmd)
        .arg("generate")
        .arg(backend)
        .arg(make)
        .pass_stdin("one")
        .pass_stdin("two");

    // Verify and cleanup
    env.finish().expect("temp folder not empty at end of test");
    assert_cmd_snapshot!(cmd);
}

#[test]
fn generator_failes_scenario() {
    let root = project_root();
    let backend = root.join("examples/generator_failes/backend.toml");
    let make = root.join("examples/generator_failes/make.json");

    let env = TempTestEnv::new();

    let mut cmd = sdtin_cli("one\ntwo\n");

    env.apply_env(&mut cmd)
        .arg("generate")
        .arg(backend)
        .arg(make)
        .pass_stdin("one")
        .pass_stdin("two");

    // Verify and cleanup
    env.finish().expect("temp folder not empty at end of test");

    assert_cmd_snapshot!(cmd);
}

#[test]
fn scenario_help() {
    let mut cmd = cli();
    cmd.arg("generate").arg("--help");

    assert_cmd_snapshot!(cmd);
}
