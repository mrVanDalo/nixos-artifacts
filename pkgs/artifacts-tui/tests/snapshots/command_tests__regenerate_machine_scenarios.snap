---
source: tests/command_tests.rs
info:
  program: artifacts
  args:
    - "--log-level=debug"
    - regenerate
    - /home/palo/dev/artifacts/nixos-artifacts/pkgs/artifacts-tui/examples/bigger_setup
    - "--machine=machine-one"
  env:
    NIXOS_ARTIFACTS_BACKEND_CONFIG: /home/palo/dev/artifacts/nixos-artifacts/pkgs/artifacts-tui/examples/bigger_setup/backend.toml
    TMPDIR: /tmp/artifacts-tui-ci
  stdin: ""
---
success: true
exit_code: 0
----- stdout -----
DEBUG: Running nix build: "/run/current-system/sw/bin/nix" "build" "--impure" "-I" "flake=/home/palo/dev/artifacts/nixos-artifacts/pkgs/artifacts-tui/examples/bigger_setup" "--no-link" "--print-out-paths" "--expr" "\nlet\n  system = \"x86_64-linux\";\n  filterAttrs =\n    pred: set:\n    builtins.removeAttrs set (builtins.filter (name: !pred name set.${name}) (builtins.attrNames set));\n  flake = builtins.getFlake (toString <flake>);\n  pkgs = flake.inputs.nixpkgs.legacyPackages.${system};\n  configurations = builtins.attrNames (\n    filterAttrs (\n      machine: configuration: builtins.hasAttr \"artifacts\" configuration.options\n    ) flake.nixosConfigurations\n  );\n  make = map (name: {\n    machine = name;\n    artifacts = flake.nixosConfigurations.${name}.config.artifacts.store;\n    config =\n      if (builtins.hasAttr \"config\" flake.nixosConfigurations.${name}.config.artifacts) then\n        flake.nixosConfigurations.${name}.config.artifacts.config\n      else\n        { };\n  }) configurations;\nin\npkgs.writeText \"test.json\" (builtins.toJSON make)\n"
DEBUG: make config: [{"artifacts":{"artifact-one":{"files":{"first-secret":{"group":"root","mode":"0400","name":"first-secret","owner":"root","path":"/run/secrets/first-secret"}},"generator":"/nix/store/y3nn4hywc8n4ymlzi5h7l2qysk5icl3v-test_generator_one.sh","name":"artifact-one","prompts":{},"serialization":"test","shared":false},"artifact-two":{"files":{"second-secret":{"group":"root","mode":"0400","name":"second-secret","owner":"root","path":"/run/secrets/second-secret"}},"generator":"/nix/store/dvr437j28bzg11ymlj0kfcpifazkj3s7-test_generator_two.sh","name":"artifact-two","prompts":{},"serialization":"test","shared":false}},"config":{},"machine":"machine-one"},{"artifacts":{"artifact-one":{"files":{"first-secret":{"group":"root","mode":"0400","name":"first-secret","owner":"root","path":"/run/secrets/first-secret"}},"generator":"/nix/store/y3nn4hywc8n4ymlzi5h7l2qysk5icl3v-test_generator_one.sh","name":"artifact-one","prompts":{},"serialization":"test","shared":false},"artifact-two":{"files":{"second-secret":{"group":"root","mode":"0400","name":"second-secret","owner":"root","path":"/run/secrets/second-secret"}},"generator":"/nix/store/dvr437j28bzg11ymlj0kfcpifazkj3s7-test_generator_two.sh","name":"artifact-two","prompts":{},"serialization":"test","shared":false}},"config":{},"machine":"machine-two"}]
[regenerate]
DEBUG:     files to produce -> 1 files
DEBUG:       - first-secret => /run/secrets/first-secret owner=root group=root
âš¡ machine-one/artifact-one
DEBUG: bwrap command: bwrap --unshare-all --unshare-user --uid 1000 --gid 1000 --tmpfs / --chdir / --ro-bind /nix/store /nix/store --tmpfs /usr/lib/systemd --proc /proc --dev /dev --bind /tmp/artifacts-tui-ci/prompt-artifact-one /tmp/artifacts-tui-ci/prompt-artifact-one --bind /tmp/artifacts-tui-ci/out-artifact-one /tmp/artifacts-tui-ci/out-artifact-one --ro-bind /nix/store/afn948adxz0x9hxhz6b5kzf09a14jdqx-test_generator_one.sh/bin /nix/store/afn948adxz0x9hxhz6b5kzf09a14jdqx-test_generator_one.sh/bin --ro-bind /bin /bin --ro-bind /usr/bin /usr/bin --ro-bind /tmp/artifacts-tui-ci/artifacts-tui-passwd-371ed15b95352dfc.txt /etc/passwd -- /bin/sh /nix/store/afn948adxz0x9hxhz6b5kzf09a14jdqx-test_generator_one.sh/bin/test_generator_one.sh
ðŸ’¾ serialize secrets
=== Content of /tmp/artifacts-tui-ci/out-artifact-one/first-secret ===
one
=========================
[regenerate]
DEBUG:     files to produce -> 1 files
DEBUG:       - second-secret => /run/secrets/second-secret owner=root group=root
âš¡ machine-one/artifact-two
DEBUG: bwrap command: bwrap --unshare-all --unshare-user --uid 1000 --gid 1000 --tmpfs / --chdir / --ro-bind /nix/store /nix/store --tmpfs /usr/lib/systemd --proc /proc --dev /dev --bind /tmp/artifacts-tui-ci/prompt-artifact-two /tmp/artifacts-tui-ci/prompt-artifact-two --bind /tmp/artifacts-tui-ci/out-artifact-two /tmp/artifacts-tui-ci/out-artifact-two --ro-bind /nix/store/yr229cqf9dxg8sffmk8pbd37wwza5988-test_generator_two.sh/bin /nix/store/yr229cqf9dxg8sffmk8pbd37wwza5988-test_generator_two.sh/bin --ro-bind /bin /bin --ro-bind /usr/bin /usr/bin --ro-bind /tmp/artifacts-tui-ci/artifacts-tui-passwd-ebde385bfb41d4e2.txt /etc/passwd -- /bin/sh /nix/store/yr229cqf9dxg8sffmk8pbd37wwza5988-test_generator_two.sh/bin/test_generator_two.sh
ðŸ’¾ serialize secrets
=== Content of /tmp/artifacts-tui-ci/out-artifact-two/second-secret ===
two
=========================
âœ… machine-two/artifact-one
âœ… machine-two/artifact-two

----- stderr -----
